src/library_html5.js

```diff
@@ -2598,6 +2598,157 @@ var LibraryJSEvents = {
 
   emscripten_html5_remove_all_event_listeners: function() {
     JSEvents.removeAllEventListeners();
+  },
+
+  emscripten_request_animation_frame: function(cb, userData) {
+    return requestAnimationFrame(function(timeStamp) {
+      Module['dynCall_idi'](cb, timeStamp, userData);
+    });
+  },
+
+  emscripten_cancel_animation_frame: function(id) {
+    cancelAnimationFrame(id);
+  },
+
+  emscripten_request_animation_frame_loop: function(cb, userData) {
+    function tick(timeStamp) {
+      if (Module['dynCall_idi'](cb, timeStamp, userData)) requestAnimationFrame(tick);
+    }
+    return requestAnimationFrame(tick);
+  },
+
+  _polyfill_set_immediate__postset:
+    'var __setImmediate_id_counter = 0;\n' +
+    'var __setImmediate_queue = [];\n' +
+    'var __setImmediate_message_id = "_si";\n' +
+    'function __setImmediate_cb(e) {\n' +
+      'if (e.data === __setImmediate_message_id) {\n' +
+        'e.stopPropagation();\n' +
+        '__setImmediate_queue.shift()();\n' +
+        '++__setImmediate_id_counter;\n' +
+      '}\n' +
+    '}\n' +
+    'if (typeof setImmediate === "undefined") {\n' +
+      'addEventListener("message", __setImmediate_cb, true);\n' +
+      'setImmediate = function(func) {\n' +
+        'postMessage(__setImmediate_message_id, "*");\n' +
+        'return __setImmediate_id_counter + __setImmediate_queue.push(func) - 1;\n' +
+      '}\n' +
+      'clearImmediate = function(id) {\n' +
+        'var index = id - __setImmediate_id_counter;\n' +
+        'if (index >= 0 && index < __setImmediate_queue.length) __setImmediate_queue[index] = function(){};\n' + // must preserve the order and count of elements in the queue, so replace the pending callback with an empty function
+      '}\n' +
+    '}',
+
+  _polyfill_set_immediate: function() { /* nop, used for its postset to ensure setImmediate() polyfill is not duplicated between emscripten_set_immediate() and emscripten_set_immediate_loop().*/ },
```

Comment:
why is the problem described by the comment a problem? if two functions use a third, the third will be included just once?

Reply:
It is not about the functions using a third function, but that both `emscripten_set_immediate()` and `emscripten_setimmediate_loop()` call `setImmediate()`, so they both need to register a polyfill for `setImmediate()`. But I can't add the polyfill registration as a `__postset` to both of those functions, since if an application might happen to reference both `emscripten_set_immediate()` and `emscripten_setimmediate_loop()`, then it would generate the polyfill registration twice. By depending on a shared dummy symbol the postset registration can be guaranteed to only appear once.

Reply:
Oh, I see. Sounds good.